lib/net.c

```diff
@@ -704,25 +701,35 @@ check_port_rss(struct gatekeeper_if *iface, unsigned int port_idx,
 	const struct rte_eth_dev_info *dev_info,
 	struct rte_eth_conf *port_conf)
 {
-	if (dev_info->flow_type_rss_offloads == 0) {
+	uint64_t rss_off = dev_info->flow_type_rss_offloads;
+
+	if (rss_off == 0) {
 		/* This port doesn't support RSS, so disable RSS. */
 		G_LOG(NOTICE, "net: port %hu (%s) on the %s interface does not support RSS\n",
 			iface->ports[port_idx], iface->pci_addrs[port_idx],
 			iface->name);
 		iface->rss = false;
-	} else if ((dev_info->flow_type_rss_offloads & GATEKEEPER_RSS_HF) !=
-			GATEKEEPER_RSS_HF) {
-		G_LOG(WARNING,
-			"net: port %hu (%s) on the %s interface only supports RSS hash functions 0x%"PRIx64", but Gatekeeper requires: 0x%"PRIx64"\n",
+	} else if ((rss_off & GATEKEEPER_RSS_HF) == 0) {
+		G_LOG(NOTICE, "net: port %hu (%s) on the %s interface does not support any IP-related RSS hashes\n",
 			iface->ports[port_idx], iface->pci_addrs[port_idx],
-			iface->name,
-			dev_info->flow_type_rss_offloads,
-			(uint64_t)GATEKEEPER_RSS_HF);
-	}
+			iface->name);
+		rss_off = 0;
+		iface->rss = false;
+	} else if ((rss_off & GATEKEEPER_RSS_HF) != GATEKEEPER_RSS_HF) {
+		G_LOG(NOTICE,
+			"net: port %hu (%s) on the %s interface only supports RSS hash functions 0x%"PRIx64", but Gatekeeper asks for 0x%"PRIx64"\n",
+			iface->ports[port_idx], iface->pci_addrs[port_idx],
+			iface->name, rss_off, (uint64_t)GATEKEEPER_RSS_HF);
 
-	port_conf->rx_adv_conf.rss_conf.rss_hf &=
-		dev_info->flow_type_rss_offloads;
+		if (!(rss_off & ETH_RSS_IPV4) || !(rss_off & ETH_RSS_IPV6)) {
+			G_LOG(WARNING, "net: port %hu (%s) on the %s interface does not support both the ETH_RSS_IPV4 and ETH_RSS_IPV6 hash functions; the device may not hash packets to the correct queues\n",
```

Comment:
1. We need to be specific which hash function is missing when only one is missing to properly inform network operators.

2. We should test if IPvX is enabled and only complain that `ETH_RSS_IPVX` is missing when it is needed. This will help network operators to better diagnose the situation.

Reply:
If instead of using a constant (i.e. `GATEKEEPER_RSS_HF`) we mask out the IPvX flags of the disabled version of IP, we can also avoid some of the warnings for the other conditions in the code.

Reply:
For the above two comments, are you referring to the case where a Gatekeeper deployment doesn't need one of IPv4 or IPv6? For example, to not warn about missing `ETH_RSS_IPV6` when the Gatekeeper interface doesn't have an IPv6 address anyway.

Reply:
Correct.